allennlp_rc/dataset_readers/transformer_squad.py

```diff
@@ -0,0 +1,325 @@
+       these in tokenized form, and the gold answer strings, accessible as ``metadata['id']``,
+       ``metadata['question']``, ``metadata['context']``, ``metadata['question_tokens']``,
+       ``metadata['context_tokens']``, and ``metadata['answers']. This is so that we can more easily use the
+       official SQuAD evaluation script to get metrics.
+
+    We also support limiting the maximum length for the question. When the context+question is too long, we run a
+    sliding window over the context and emit multiple instances for a single question. At training time, we only
+    emit instances that contain a gold answer. At test time, we emit all instances.
+
+    Parameters
+    ----------
+    transformer_model_name : ``str``, optional (default=``bert-base-cased``)
+        This reader chooses tokenizer and token indexer according to this setting.
+    length_limit : ``int``, optional (default=384)
+        We will make sure that the length of context+question never exceeds this many tokens.
+    stride : ``int``, optional (default=128)
+        When context+question are too long for the length limit, we emit multiple instances for one question,
+        where the context is shifted. This parameter specifies the overlap between the shifted context window. It
+        is called "stride" instead of "overlap" because that's what it's called in the original huggingface
+        implementation.
+    skip_invalid_examples: ``bool``, optional (default=False)
+        If this is true, we will skip examples that don't have a gold answer. You should set this to True during
+        training, and False any other time.
+    max_query_length : ``int``, optional (default=64)
+        The maximum number of wordpieces dedicated to the question. If the question is longer than this, it will be
+        truncated.
+    max_questions : ``int``, optional (default=None)
+        If this is specified, the reader quits after this many questions have been processed. This can be useful
+        for debugging.
+    """
+
+    def __init__(
+        self,
+        transformer_model_name: str = "bert-base-cased",
+        length_limit: int = 384,
+        stride: int = 128,
+        skip_invalid_examples: bool = False,
+        max_query_length: int = 64,
+        max_questions: Optional[int] = None,
+    ) -> None:
+        super().__init__()
+        self._tokenizer = PretrainedTransformerTokenizer(
+            transformer_model_name, add_special_tokens=False, calculate_character_offsets=True
+        )
+        self._token_indexers = {"tokens": PretrainedTransformerIndexer(transformer_model_name)}
+        self.length_limit = length_limit
+        self.stride = stride
+        self.skip_invalid_examples = skip_invalid_examples
+        self.max_query_length = max_query_length
+        self.max_questions = max_questions
+        if (
+            AutoModel.from_pretrained(
```

Comment:
Yes, it does that. I'm thinking we should have a way of caching the loaded transformer models, because most likely, you'll load it up again a few seconds later. @pengshuang had a thing like this in his PR. If we do this, we'll still load a huge model just to get this number, but we would have loaded the model anyways later, so it's not a problem.

The alternative is to go for a less portable solution where we just download the huggingface config and read it, without downloading the rest of the model. We'd start depending on some internal organization of those models then.

Reply:
I know we had caching for BERT models; should be easy to add something similar more generally for transformers.

Another alternative is to submit something upstream that adds an API call somewhere in a general way.  Not sure at all where that would go.